### ISSUE TITLE
Input and layer randomizations should be active only for learning function not for evaluation function

### ISSUE THREAD

**Issue by hknerdgn** [lmjohns3/theanets#65]
I think these randomizations (such as input_noise, input_dropout, hidden_dropout) are always active.

They should be disabled for evaluation functions since they are only needed for training.

This may be related to this issue mentioned here in the mailing list: https://groups.google.com/forum/#!topic/theanets/8dgkupBdugY

**Comment by lmjohns3** [comment]
Thanks for filing this! It's an interesting issue.

On the one hand, dropout and noise shouldn't be active in a "production" model, since they will only hurt performance. On the other hand, if dropout and noise are removed from the loss for validation data, then the loss computed on validation data won't be comparable to the loss computed on training data. What about other regularizers that are currently active for both validation and training data (e.g. weight_l2)? Should those be disabled as well on the validation dataset?

Also, the technical task of implementing this could require a bit of an interface change. That's not so terrible, but it will require a bit of thought.

**Comment by hknerdgn** [comment]
I am aware of some implementations that does remove the randomizations for the evaluation function in both validation and test time. I think this is most probably the right way to go. I think other regularizations should also be removed during validation and test time, since the cost function for training is irrelevant for testing. What matters is the performance which is measured by the appropriate part of the cost function, be it cross-entropy, least squares cost etc.

If random dropout is used to train the network, they usually multiply the activations of the nodes with p where p is the training dropout rate for that node. This is to match the mean value of that activation for training and test time (I think). For input dropout, no multiplication is needed since the mean of the inputs are assumed to be zero.

**Comment by lmjohns3** [comment]
This actually turned out to be a bit easier than I'd thought! I will close this for now, please reopen if you notice anything is suspicious.
